server/middleware/OktaServerSideAuthMiddleware.ts

```diff
@@ -0,0 +1,90 @@
+import { joinUrl } from '@guardian/libs';
+import type { NextFunction, Request, Response } from 'express';
+import fetch from 'node-fetch';
+import {
+	clearIdentityLocalState,
+	getIdentityLocalState,
+	setIdentityLocalState,
+} from '@/server/IdentityLocalState';
+import { OAuthAccessTokenCookieName } from '@/server/oauth';
+import { getConfig as getOktaConfig } from '@/server/oktaConfig';
+import { requiresSignin } from '../../shared/requiresSignin';
+
+declare const CYPRESS: string;
+
+type OktaUserInfo = {
+	legacy_identity_id: string;
+	email: string;
+	name: string;
+};
+
+export const withOktaSeverSideValidation = async (
+	req: Request,
+	res: Response,
+	next: NextFunction,
+) => {
+	if (CYPRESS === 'SKIP_IDAPI') {
+		return next();
+	}
+	const oktaConfig = await getOktaConfig();
+	if (!oktaConfig.useOkta) {
+		console.log('Okta disabled, skipping Okta server side validation...');
+		return next();
+	}
+
+	console.log('Validating token server side ...');
+	const signinRequired = () => requiresSignin(req.originalUrl);
+
+	const locallyValidatedIdentityData = getIdentityLocalState(res);
+	const accessToken = req.signedCookies[OAuthAccessTokenCookieName];
+	if (!accessToken || !locallyValidatedIdentityData?.userId) {
+		if (signinRequired()) {
+			console.error(
+				'error: no access token or user in request for a sign-in required endpoint! this should have failed local validation',
+			);
+			return res.sendStatus(500);
+		} else {
+			return next();
+		}
+	}
+
+	const issuerUrl = joinUrl(
+		oktaConfig.orgUrl,
+		'/oauth2/',
+		oktaConfig.authServerId,
+	);
+
+	const oktaResponse = await fetch(`${issuerUrl}/v1/userinfo/`, {
+		method: 'GET',
+		headers: {
+			'Content-Type': 'application/json',
+			Authorization: `Bearer ${accessToken}`,
+		},
+	});
+
+	console.log(`okta response status: ${oktaResponse.status}`);
+	if (oktaResponse.status == 200) {
+		//valid token
+		const oktaUserInfo = await oktaResponse.json<OktaUserInfo>();
+		setIdentityLocalState(res, {
+			signInStatus: 'signedInRecently', // TODO can I hardcode this here or do I need to check something else ?
+			userId: oktaUserInfo.legacy_identity_id,
+			name: oktaUserInfo.name,
+			email: oktaUserInfo.email,
+		});
+		return next();
+	} else if ([403, 401].includes(oktaResponse.status)) {
+		//invalid token
+		clearIdentityLocalState(res);
+		if (signinRequired()) {
+			return res.sendStatus(401);
+		}
+		return next();
+	} else {
+		//unexpected return status
+		console.error(
+			`unexpected status from okta userinfo ${oktaResponse.status}`,
+		);
+		return res.sendStatus(500);
+	}
```

Comment:
If you wanted to turn the fetch into a method as we were discussing, you could do something like this:

```javascript

interface OktaValidationResponse {
	ok: boolean;
	valid: boolean;
	userId?: string;
	name?: string;
	email?: string;
}

const validateInOkta = async ({
	oktaConfig,
	accessToken
}: {
	oktaConfig: OktaConfig;
	accessToken: string;
}): OktaValidationResponse => {
	const issuerUrl = joinUrl(
		oktaConfig.orgUrl,
		'/oauth2/',
		oktaConfig.authServerId,
	);

	try {
		const oktaResponse = await fetch(`${issuerUrl}/v1/userinfo/`, {
			method: 'GET',
			headers: {
				'Content-Type': 'application/json',
				Authorization: `Bearer ${accessToken}`,
			},
		});
		if (oktaResponse.status === 200) {
			//valid token
			const oktaUserInfo = await oktaResponse.json<OktaUserInfo>();
			return {
				ok: true,
				valid: true,
				userId: oktaUserInfo.legacy_identity_id,
				name: oktaUserInfo.name,
				email: oktaUserInfo.email,
			};
		}  else if ([403, 401].includes(oktaResponse.status)) {
			//invalid token
			return { ok: true, valid: false };
		} else {
			//unexpected return status
			return { ok: false, valid: false };
		}
	} catch (error) {
		console.error(error);
		return { ok: false, valid: false };
	}
}

const oktaValidationResponse = await validateInOkta({
	oktaConfig,
	accessToken,
});

if (!oktaValidationResponse.ok) {
	return res.sendStatus(500);
}

if (!oktaValidationResponse.valid) {
	return res.sendStatus(401);
}

return next();
```